internal/sqlsmith: TestGenerateParse failed

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18714167?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18714167?buildTab=artifacts#/) on master @ [ef0b34c33974d515f0d6f82367a781ac10ce3653](https://github.com/cockroachdb/cockroach/commits/ef0b34c33974d515f0d6f82367a781ac10ce3653):


```
	*
FROM
	(
		SELECT
			"t%p 😚ab_159".col_204 AS col_206,
			"t%p 😚ab_159".col_205 AS col_207,
			"t%p 😚ab_159"."coL_203" AS col_208
		FROM
			(
				VALUES
					(NULL, COALESCE(NULL, NULL), similar_escape('G':::STRING::STRING)::STRING COLLATE en_US),
					(NULL, NULL, NULL),
					(NULL, NULL, NULL)
			)
				AS "t%p 😚ab_159" ("coL_203", col_204, col_205)
		WHERE
			'A':::STRING !~ 'C':::STRING
	)
		AS tab_160
ORDER BY
	col_206 ASC NULLS FIRST, col_207 DESC, col_208 NULLS FIRST;

STMT: 90
SELECT
	*
FROM
	(
		SELECT
			tab_161."col%v_209" AS "%vcol_210"
		FROM
			(
				VALUES
					('00:08:00':::TIME),
					('00:07:00':::TIME),
					('00:02:00':::TIME),
					('00:02:00':::TIME),
					('00:00:00':::TIME)
			)
				AS tab_161 ("col%v_209")
		ORDER BY
			tab_161."col%v_209" DESC
	)
		AS tab_162
ORDER BY
	"%vcol_210" ASC NULLS LAST;

    setup_test.go:116: SET SESSION CHARACTERISTICS AS TRANSACTION: at or near "EOF": syntax error
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f82616d5c9a8c3209014f997587365ae/logTestGenerateParse1486468227
--- FAIL: TestGenerateParse (4.14s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47149

--- a/pkg/internal/sqlsmith/relational.go
+++ b/pkg/internal/sqlsmith/relational.go
@@ -1471,7 +1471,7 @@ func makeInsert(s *Smither) (tree.Statement, bool) {
 	return stmt, ok
 }
 
-func makeTransactionModes(s *Smither) (tree.TransactionModes, bool) {
+func makeTransactionModes(s *Smither) tree.TransactionModes {
 	levels := []tree.IsolationLevel{
 		tree.UnspecifiedIsolation,
 		tree.ReadUncommittedIsolation,
@@ -1480,12 +1480,11 @@ func makeTransactionModes(s *Smither) (tree.TransactionModes, bool) {
 		tree.SnapshotIsolation,
 		tree.SerializableIsolation,
 	}
-	modes := tree.TransactionModes{Isolation: levels[s.rnd.Intn(len(levels))]}
-	return modes, true
+	return tree.TransactionModes{Isolation: levels[s.rnd.Intn(len(levels))]}
 }
 
 func makeBegin(s *Smither) (tree.Statement, bool) {
-	modes, _ := makeTransactionModes(s)
+	modes := makeTransactionModes(s)
 	return &tree.BeginTransaction{Modes: modes}, true
 }
 
@@ -1531,7 +1530,14 @@ func makeSetSessionCharacteristics(s *Smither) (tree.Statement, bool) {
 	if s.disableIsolationChange {
 		return nil, false
 	}
-	modes, _ := makeTransactionModes(s)
+	modes := makeTransactionModes(s)
+	fmtCtx := tree.NewFmtCtx(tree.FmtSimple)
+	modes.Format(fmtCtx)
+	if fmtCtx.String() == "" {
+		// If no options are specified, then we would get an invalid SET SESSION
+		// CHARACTERISTICS statement.
+		return nil, false
+	}
 	return &tree.SetSessionCharacteristics{Modes: modes}, true
 }
 

